core/src/main/java/io/questdb/cairo/map/UnorderedVarcharMap.java

```diff
@@ -0,0 +1,740 @@
+                }
+            }
+        }
+    }
+
+    private void rehash() {
+        rehash((long) keyCapacity << 1);
+    }
+
+    private void rehash(long newKeyCapacity) {
+        if (nResizes == maxResizes) {
+            throw LimitOverflowException.instance().put("limit of ").put(maxResizes).put(" resizes exceeded in unordered map");
+        }
+        if (newKeyCapacity > MAX_SAFE_INT_POW_2) {
+            throw CairoException.nonCritical().put("map capacity overflow");
+        }
+        if (newKeyCapacity <= keyCapacity) {
+            return;
+        }
+
+        final long newSizeBytes = entrySize * newKeyCapacity;
+        final long newMemStart = Unsafe.malloc(newSizeBytes, memoryTag);
+        final long newMemLimit = newMemStart + newSizeBytes;
+        Vect.memset(newMemStart, newSizeBytes, 0);
+        final int newMask = (int) newKeyCapacity - 1;
+
+        for (long addr = memStart; addr < memLimit; addr += entrySize) {
+            long packedHashSizeFlags = Unsafe.getUnsafe().getLong(addr);
+            if (packedHashSizeFlags == 0) {
+                continue;
+            }
+
+            int hash = Unsafe.getUnsafe().getInt(addr);
+            long newAddr = getStartAddress(newMemStart, hash & newMask);
+            while (Unsafe.getUnsafe().getLong(newAddr) != 0) {
+                newAddr += entrySize;
+                if (newAddr >= newMemLimit) {
+                    newAddr = newMemStart;
+                }
+            }
+            Vect.memcpy(newAddr, addr, entrySize);
+        }
+
+        Unsafe.free(memStart, memLimit - memStart, memoryTag);
+
+        memStart = newMemStart;
+        memLimit = newMemStart + newSizeBytes;
+        mask = newMask;
+        free += (int) ((newKeyCapacity - keyCapacity) * loadFactor);
+        keyCapacity = (int) newKeyCapacity;
+        nResizes++;
+    }
+
+    private UnorderedVarcharMapValue valueOf(long startAddress, boolean newValue, UnorderedVarcharMapValue value) {
+        return value.of(startAddress, memLimit, newValue);
+    }
+
+    static boolean isAscii(byte flags) {
+        return (flags & FLAG_IS_ASCII) != 0;
+    }
+
+    static boolean isNull(byte flags) {
+        return (flags & FLAG_IS_NULL) != 0;
+    }
+
+    static boolean isSizeNull(int sizeAndFlags) {
+        return (sizeAndFlags & SIZE_IS_NULL) != 0;
+    }
+
+
+    static byte unpackFlags(long packedHashSizeFlags) {
+        return (byte) (packedHashSizeFlags >>> 56);
+    }
+
+    static int unpackSize(long packedHashSizeFlags) {
+        int sizeAndFlags = (int) (packedHashSizeFlags >>> 32);
+
+        // clear top 2 bits
+        return sizeAndFlags & 0x3FFFFFFF;
+    }
+
+    long entrySize() {
+        return entrySize;
+    }
+
+    boolean isZeroKey(long startAddress) {
+        long packedHashAndSize = Unsafe.getUnsafe().getLong(startAddress);
+        return packedHashAndSize == 0;
+    }
+
+    class Key implements MapKey {
+        private byte flags;
+        private long ptr;
```

Comment:
nit: do we need this field considering that applying mask to `ptrWithUnstableFlag` is cheap?

Reply:
this was meant to be a correctness thing rather than a performance thing. to make it clear where you do have to mask out the flag and where you don't have to. but I can look how it would look with just `ptrWithUnstableFlags`

Reply:
@puzpuzpuz:

I tried it here: https://github.com/questdb/questdb/pull/4435/commits/2139a00fb3acd60090bae33bf91442a636c26462

but I am not convinced it's an improvement. for example see this:
```java
        @Override
        public MapValue createValue() {
            long hash = Hash.hashMem64(ptrWithUnstableFlag & PTR_MASK, size);
            return createValue(hash);
        }
```
we calculate the plain pointer to calculate a hash. and then we throw it away. only it calculate it later inside `createValue()`. 
I could optimize this by passing the plain ptr to `createValue()`. but would that be an improvement compared to the version with 2 fields? I am not sure. WDYT?

Reply:
A bitwise and is much less expensive than a hash function call, so I wouldn't worry about having it. On the other hand, I find these two `ptr*` confusing.